src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,373 @@
+        where TExpressionSyntax : SyntaxNode
+        where TMethodDeclarationSyntax : SyntaxNode
+        where TInvocationExpressionSyntax : SyntaxNode
+    {
+        protected const string ExpressionAnnotationKind = nameof(ExpressionAnnotationKind);
+        protected abstract Task<Solution> IntroduceParameterAsync(SemanticDocument document, TExpressionSyntax expression, bool allOccurrences, bool trampoline, CancellationToken cancellationToken);
+        protected abstract bool ExpressionWithinParameterizedMethod(TExpressionSyntax expression);
+        protected abstract Task<Solution> RewriteCallSitesAsync(TExpressionSyntax expression, ImmutableDictionary<Document, List<TInvocationExpressionSyntax>> callSites, IMethodSymbol methodSymbol, CancellationToken cancellationToken);
+
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var (document, textSpan, cancellationToken) = context;
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var action = await IntroduceParameterAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (action != null)
+            {
+                context.RegisterRefactoring(action, textSpan);
+            }
+        }
+
+        public async Task<CodeAction> IntroduceParameterAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var semanticDocument = await SemanticDocument.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+
+            var expression = await document.TryGetRelevantNodeAsync<TExpressionSyntax>(textSpan, cancellationToken).ConfigureAwait(false);
+            if (expression == null || CodeRefactoringHelpers.IsNodeUnderselected(expression, textSpan))
+            {
+                return null;
+            }
+
+            var expressionType = semanticDocument.SemanticModel.GetTypeInfo(expression, cancellationToken).Type;
+            if (expressionType is IErrorTypeSymbol)
+            {
+                return null;
+            }
+
+            var containingType = expression.AncestorsAndSelf()
+                .Select(n => semanticDocument.SemanticModel.GetDeclaredSymbol(n, cancellationToken))
+                .OfType<INamedTypeSymbol>()
+                .FirstOrDefault();
+
+            containingType ??= semanticDocument.SemanticModel.Compilation.ScriptClass;
+
+            if (containingType == null || containingType.TypeKind == TypeKind.Interface)
+            {
+                return null;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            if (expression != null)
```

Comment:
this check should be done first before all this expensive semantic work.

Reply:
moved